### ISSUE TITLE
Static Files Truncated at 65,339 or 65340 via TLS connection

### ISSUE THREAD

**Issue by daninus14** [fukamachi/woo#120]
In Chrome I am getting the following error for _some_ Static Files from the server

```
net::ERR_CONTENT_LENGTH_MISMATCH 200 (OK)

```

The headers seem to be alright, however the actual files are being truncated

```
connection:    keep-alive
content-length:    162770
content-type:    text/css; charset=utf-8
```

The actual files being transferred are truncated partially and only have 65,339 or 65340 characters in length.

I just tried changing back to Hunchentoot with the caveman2 app and it works. Actually I just checked and interestingly enough the actual length of characters in the file is 162762 but the `content-length` header has 162770. It could be some escape characters that the browser just renders. Unclear. Nevertheless, this works with Hunchentoot and not with woo when removing `:server :woo`.

It looks like somewhere in woo it's limiting the file size to be sent over to 65,339 or 65,340. Just want to point out that `2^16 = 65536`. So I imagine somewhere in the code there must be an integer handling how much to actually transfer, and because it's maxed out, it truncates the transfer. Probably changing this to a larger integer of say 32 bits is not such a good idea, because any file longer than that (I think about 4 GB) will then just be truncated as well. I wonder where the `content-length` header is being set, from reading the file size in the filesystem probably, and it could make sense to look into that code as a sample to dealing with the integers. Setting it to an integer 64 bits long will probably solve the issue for a long time though.

I'll be happy to help if you can point out to me which files and functions are dealing with the file transfer.

**Comment by fukamachi** [comment]
Thank you for reporting!

Is this only when SSL is enabled, or always?
It may be because of a new function [wev:write-socket-stream](https://github.com/fukamachi/woo/blob/7f5219c55d49190f5ae17b123a8729b31c5d706e/src/ev/socket.lisp#L170) which sends a static file.

**Comment by daninus14** [comment]
Hi, finally got around to checking it. I can confirm this only happens when SSL is enabled.

I see here is the place where that is used: https://github.com/fukamachi/woo/blob/7f5219c55d49190f5ae17b123a8729b31c5d706e/src/woo.lisp#L391

And just below seems to be without SSL https://github.com/fukamachi/woo/blob/7f5219c55d49190f5ae17b123a8729b31c5d706e/src/woo.lisp#L405

It looks like the non SSL version is using this function: https://github.com/fukamachi/woo/blob/7f5219c55d49190f5ae17b123a8729b31c5d706e/src/ev/socket.lisp#L357

Not sure how that last function is actually sending the file, just by doing setf on `sendfile-fd`? Why can't we do the same by SSL? What about `write-socket-stream` is unique to SSL that it was needed to begin with?

**Comment by fukamachi** [comment]
Thank you for your efforts to look into it, but I still couldn't reproduce it.
Woo can serve a 147KB PNG file and a 385 KB YAML file via HTTPS on my machine.
It can be an environmental issue.

> Not sure how that last function is actually sending the file, just by doing setf on sendfile-fd? Why can't we do the same by SSL? What about write-socket-stream is unique to SSL that it was needed to begin with?

Because a system call `sendfile` can't send a file over a TLS connection by encrypting it.
The latest Linux supports kernel-level TLS, but it's off by default and we need to build it on our own to use it.

**Comment by daninus14** [comment]
Thanks, I'm just using a fresh ubuntu LTS without nginx or any reverse proxy, just straight woo. I guess if it comes up for anyone else they may help us figure it out

**Comment by solipsismes** [comment]
I think I have a reproduction, and I think I know why it hasn't reproduced for anyone else: **it needs a slow client.**

A fast localhost client drains the socket as fast as the server fills it, so the teardown never races data still in flight. Serving ten file sizes from 1KB to 2MB with `dexador` over HTTPS on my machine, every one transfers correctly — same as @fukamachi's PNG and YAML. The bug only shows when the reader lags behind the writer, which is exactly what a browser over a real network does.

## Reproducing

Serve a file as a pathname body over HTTPS, then connect with a client that requests it, stalls a few seconds, and only then drains slowly:

```
file on disk:         162770 bytes
received (incl hdrs): 159744 bytes   <-- short by ~3KB
client error: SSL-ERROR-SSL: error:0A000126:SSL routines::
              unexpected eof while reading (rec_layer_s3.c:703)
```

Deterministic — identical byte count across runs. The same slow client against **plain HTTP** transfers in full (162890 = 162770 + 120 header bytes), which matches @daninus14's report that this is SSL-only.

## What seems to be happening

It isn't a write bug. I instrumented `flush-buffer`, and `ssl-write` reports writing everything:

```
[flush-buffer] ssl=t len=162890 n=162890
```

The server writes every byte. The problem is the teardown: `close-socket` calls `close(2)` on the fd directly, and `SSL_shutdown` doesn't appear anywhere in `src/` — so the connection ends without a TLS `close_notify`. The peer sees the stream end mid-record, OpenSSL 3.x treats that as `unexpected eof while reading` rather than a clean close, and data still in flight is discarded. From the browser's side that's a response shorter than `Content-Length`, i.e. `ERR_CONTENT_LENGTH_MISMATCH`.

Adding an `SSL_shutdown` before the `close(2)` fixes it here:

```
before: 159744 / 162890   client: "unexpected eof while reading"
after:  162890 / 162890   client: clean EOF
```

The existing test suite still passes (HTTP + SSL + IPv6).

Tested on SBCL 2.6.6, macOS arm64, OpenSSL 3.6.3.

## What I have not established

**I'm not certain this is the same failure @daninus14 hit.** Mine loses ~3KB of 162KB; theirs truncated at 65,339 of 162,770 — same class, very different magnitude, and I can't explain the 65,339 constant. It's possible the missing `close_notify` is one of several ways this breaks and their window was much larger (Linux, real network, Chrome), or that theirs is something else entirely.

Two other things worth flagging, which I have not fixed:

- `SSL_shutdown` can return `-1`/`WANT_WRITE` on a non-blocking socket if `close_notify` can't be sent immediately. Calling it and closing anyway still drops the notify under write pressure; a complete fix would defer the close until it drains.
- The SSL handle looks like it leaks: `cl+ssl`'s `with-new-ssl` only frees on abort, so Woo owns the handle after `init-ssl-handle` returns, and `close-socket` never calls `SSL_free`. Separate issue.

Happy to open a PR with the `close_notify` change if it's useful, and happy to share the repro script.

@daninus14 — if you can still reproduce your original case, it would be interesting to know whether an `SSL_shutdown` in `close-socket` changes anything on your setup, since your truncation is much larger than mine.

**Comment by solipsismes** [comment]
Correction to my last comment: `close_notify` is real, but it's the smaller of two bugs here. I've found the main one, and **you can now reproduce it on your machine in one command.**

## Why it never reproduced

`SO_SNDBUF`. On macOS loopback the send buffer auto-tunes large enough that a whole response fits in a single `SSL_write`, so the failing path is unreachable. Force the send buffer down to a Linux-ish size on the accepted socket and it fires immediately, every time. A 147KB PNG over loopback genuinely can't fail here — it isn't an environmental fluke, it's structural.

## What actually happens

With `SO_SNDBUF` forced to 8192 and a client that lags behind the writer:

```
[flush-buffer] len=162890 n=-1 sap=#X80081E0010  -> errno 3 (SSL_ERROR_WANT_WRITE)
[flush-buffer] len=162890 n=-1 sap=#X8008210010  -> errno 1 (SSL_ERROR_SSL)
```

1. The response outgrows the send buffer. `SSL_write` is all-or-nothing by default, so instead of a partial count it returns `WANT_WRITE`.
2. `flush-buffer`'s `WANT_WRITE` branch is empty — nothing is re-queued, so the caller just retries.
3. `finish-output-buffer` + `with-pointer-to-vector-data` re-pin the data at a **different address** on the retry (the two `sap` values above).
4. OpenSSL requires a retried `SSL_write` to use the same address unless `SSL_MODE_ACCEPT_MOVING_WRITE_BUFFER` is set. It isn't, so the retry is a fatal `SSL_ERROR_SSL` and the connection dies mid-response.

The client keeps whatever reached it before the failure — which lands near the send buffer size. That's why the truncation looks like it's pinned to a ~64KB boundary.

The plain-HTTP path is unaffected: `write()` reports partial counts natively and has no moving-buffer constraint, which is why this is HTTPS-only.

## Fix

Set both modes on the handle at creation (`SSL_set_mode` is a C macro over `SSL_ctrl`; cl+ssl exposes neither it nor the constants, so they're defined locally):

- `ENABLE_PARTIAL_WRITE` — `SSL_write` reports what it accepted, so the partial-write path already in `flush-buffer` (`reset-buffer` + `fast-write-sequence` from offset `n`) re-queues the remainder, exactly as the `write()` path does.
- `ACCEPT_MOVING_WRITE_BUFFER` — makes the re-pinned retry legal.

Measured with a slow TLS client, serving a 162770-byte file (162890 with headers):

| `SO_SNDBUF` | before | after |
|---|---|---|
| 8192 | 0 | 162890 |
| 32768 | 16384 | 162890 |
| 65536 | 49152 | 162890 |

The `close_notify` change from my previous comment is still needed on top — without it the last few KB are still lost to the abrupt teardown. Together they give a full transfer at every buffer size I tried. Existing suite passes.

## What I still can't account for

I can't produce @daninus14's exact 65,339. The truncation point tracks `SO_SNDBUF` and the client's read pattern, so it's environment-specific — I get 49152 at `SO_SNDBUF=65536`, not 65339. The mechanism explains *why* it sits near a 64KB boundary, but I'm not claiming to have derived their number, and I'd rather say so than fit a story to it.

@daninus14 — if you can still reproduce, this is the thing to test. On Linux the default send buffer should make it fire without any special setup.

I have both changes on a branch and I'm happy to open a PR, or leave them here if you'd rather take a different approach — in particular, setting the modes on the context rather than per-handle, or fixing `flush-buffer`'s `WANT_WRITE` branch to re-queue instead of relying on partial writes, are both reasonable alternatives I didn't pursue.

**Comment by daninus14** [comment]
Hi David @solipsismes , 

The error for me was only happening in a server. When I ran woo locally, everything worked. When I deployed to the server it broke. In particular only when using the SSL option as true. 

Right now I'm using woo without SSL directly and just using a reverse proxy server to deal with the encryption. 

What exactly is the change you want me to test? Is it the C SSL Library's code?

I saw you have a PR: https://github.com/fukamachi/woo/pull/129 Is that the fix you want me to test?

**Comment by solipsismes** [comment]
Hi @daninus14 — no, #129 isn't it. That one fixes an unrelated data race in file-size reporting; it happened to be the PR you found, but it won't touch this.

The change for your bug is in Woo's Lisp SSL code, not the C library. I've just opened it as **#131**. It (1) sets two `SSL_write` modes so a partial/retried write is handled correctly, and (2) sends a TLS `close_notify` before closing the socket so the tail of the response isn't dropped.

Your report lines up exactly with what I found: it only bites on a real server over SSL, never on loopback, because it depends on the send buffer filling up mid-response. Local transfers are small/fast enough to never trip it.

If you can re-enable SSL in your server setup and serve a file bigger than ~64KB against #131, that'd confirm it. Totally understand if the reverse-proxy setup is working for you now — no rush.
